packages/opentelemetry-plugin-http/src/http.ts

```diff
@@ -0,0 +1,437 @@
+          }
+
+          span.end();
+        });
+        HttpPlugin.setSpanOnError(span, response);
+      });
+
+      HttpPlugin.setSpanOnError(span, request);
+
+      plugin._logger.debug('makeRequestTrace return request');
+      return request;
+    };
+  }
+
+  private incomingRequestFunction(original: (event: string, ...args: unknown[]) => boolean, plugin: HttpPlugin) {
+    return function incomingRequest(this: {}, event: string, ...args: unknown[]): boolean {
+      // Only traces request events
+      if (event !== 'request') {
+        return original.apply(this, [event, ...args]);
+      }
+
+      const request = args[0] as IncomingMessage;
+      const response = args[1] as ServerResponse;
+      const path = request.url ? url.parse(request.url).pathname || '' : '';
+      const method = request.method || 'GET';
+      plugin._logger.debug('%s plugin incomingRequest', plugin.moduleName);
+
+      if (HttpPlugin.isIgnored(path, request, plugin.options.ignoreIncomingPaths)) {
+        return original.apply(this, [event, ...args]);
+      }
+
+      const propagation = plugin._tracer.getHttpTextFormat();
+      const headers = request.headers;
+
+      const spanOptions: SpanOptions = {
+        kind: SpanKind.SERVER
+      };
+
+      const spanContext = propagation.extract(HttpPlugin.PROPAGATION_FORMAT, headers);
+      if (spanContext && isValid(spanContext)) {
+        spanOptions.parent = spanContext;
+      }
+
+      const rootSpan = plugin._tracer.startSpan(path, spanOptions);
+      return plugin._tracer.withSpan(rootSpan, () => {
+        plugin._tracer.scopeManager.bind(request);
+        plugin._tracer.scopeManager.bind(response);
+
+        // Wraps end (inspired by:
+        // https://github.com/GoogleCloudPlatform/cloud-trace-nodejs/blob/master/src/plugins/plugin-connect.ts#L75)
+        const originalEnd = response.end;
+        response.end = function(this: ServerResponse, ...args: ResponseEndArgs) {
+          response.end = originalEnd;
+          // Cannot pass args of type ResponseEndArgs, Expected 1-2 arguments, but got 1 or more.
+          // tslint:disable-next-line:no-any
+          const returned = response.end.apply(this, arguments as any);
+          const requestUrl = request.url ? url.parse(request.url) : null;
+          const host = headers.host || 'localhost';
+          const userAgent = (headers['user-agent'] || headers['User-Agent']) as string;
+
+          rootSpan.setAttribute(HttpPlugin.ATTRIBUTE_HTTP_HOST, host.replace(/^(.*)(\:[0-9]{1,5})/, '$1'));
+
+          rootSpan.setAttribute(HttpPlugin.ATTRIBUTE_HTTP_METHOD, method);
+          if (requestUrl) {
+            rootSpan.setAttribute(HttpPlugin.ATTRIBUTE_HTTP_PATH, requestUrl.pathname || '');
+            rootSpan.setAttribute(HttpPlugin.ATTRIBUTE_HTTP_ROUTE, requestUrl.path || '');
+          }
+          if (userAgent) {
+            rootSpan.setAttribute(HttpPlugin.ATTRIBUTE_HTTP_USER_AGENT, userAgent);
```

Comment:
Do we need to use `.toString()` here, same as 
https://github.com/open-telemetry/opentelemetry-js/pull/161/files#diff-d1b28358da758893ac27e367738e7e51R271